src/helpers/buffers/fsv.rs

```diff
@@ -0,0 +1,220 @@
+use bitvec::bitvec;
+use bitvec::prelude::BitVec;
+use std::ops::Range;
+
+/// A vector of bytes that never grows over a certain size or shrinks below that size.
+/// Vector is segmented into regions and `max_size` is a factor of number of regions.
+/// For example, if number of regions is `2`, then `max_size` must be a multiple of two.
+///
+/// Elements can be added to this vector in random order as long as the total number of elements
+/// does not exceed the overall capacity of the vector. An attempt to do so, will lead to a panic.
+/// Every element is stored by a specific offset and it is assumed that all elements are of the same
+/// size.
+///
+/// The layout of the vector with some elements added to it is presented below. In this example,
+/// vector has 3 regions and `X` indicates that space at that element is occupied.
+///
+///  region1  region2  region3
+/// [X,_,_,_][_,_,X,_][_,X,_,X]
+///
+/// Once `region1` is completely filled up, it is possible to drain the vector
+///
+/// [X,X,X,X][_,_,X,_][_,X,_,X] -> `drain` -> [_,_,X,_][_,X,_,X][_,_,_,_]
+///
+/// This vector is used inside the send buffer to keep track of messages added to it. Once first
+/// batch of messages is ready (region1 is full), it drains this vector and send those messages
+/// down to the network layer
+#[derive(Debug)]
+pub struct FixedSizeByteVec<const N: usize> {
+    data: Vec<u8>,
+    region_size: usize,
+    added: BitVec,
+    elements_drained: usize,
+}
+
+impl<const N: usize> FixedSizeByteVec<N> {
+    pub const ELEMENT_SIZE_BYTES: usize = N;
+
+    pub fn new(region_count: usize, region_size: usize) -> Self {
+        Self {
+            data: vec![0_u8; N * region_size * region_count],
+            added: bitvec![0; region_size * region_count],
+            elements_drained: 0,
+            region_size,
+        }
+    }
+
+    /// Inserts a new element to the specified position, returning the previous element at this `index`.
+    /// ## Panics
+    /// * Panics if `index` is out of bounds.
+    pub fn insert(&mut self, index: usize, elem: [u8; N]) -> Option<[u8; N]> {
```

Comment:
Why are you returning what was present at that index?

Also, it seems like you are taking an index into your structure, rather than a record number.  This code should only track a window starting from the index just past the last data that was sent, but accept a RecordId as the index.  Then you get something like:

```rust
fn insert(&mut self, i: RecordId, data: &[u8; N]) {
  let offset = usize::from(i - self.sent);
  assert!(!self.added[offset]);
  self.data[offset * N..offset*(N+1)].copy_from_slice(data);
  self.added[offset] = true;
}
```

Reply:
that is used to check that nobody inserts a record with the same record id. Taking an index rather than `RecordId`  helps to keep this struct generic and not tied up to the send buffer implementation.

Reply:
I don't care about the argument type (use `impl Into<usize>` if you prefer).  I was more concerned about this returning what was already in place, when the caller should never be putting stuff in at the same index.

Reply:
I was basically mimicking HashMap behavior here https://doc.rust-lang.org/std/collections/struct.HashMap.html#method.insert. For generic structs it seems ok to accept values put at the same index more than once. SendBuffer handles that already so I am curious what is the reason of having this check here?

> The comment suggests that the only time Some() is returned instead of None is in the case of an error. One assumes this is to help assist with debugging? If that's the case - should we return an error with this payload instead?

I think that is basically the question here. Do we consider inserting something to this data structure at the same index more than once as an error? I know for our existing use case (send buffers) it is an error but I validate it at the callsite. My assumption is that this is a generic data structure and callers need to decide what to do. 

It is easy to go either way though. I just want to make sure it is clear what I meant here and I am totally OK changing it if this argument is not strong enough